0001-dinerojs-v2.md

```diff
@@ -0,0 +1,188 @@
+
+`convert` should no longer expect a REST API and encapsulate the fetching behavior. Instead, it should expect a Promise which resolves to a `Rates` object.
+
+```js
+const rates = new Promise(resolve =>
+  resolve({
+    EUR: 0.81162
+  })
+);
+
+Dinero({ amount: 500 }).convert("EUR", { rates });
+```
+
+##### Integers & BigInts
+
+Dinero.js v1 uses integers for `amount` (through the `number` primitive), and this shouldn't change, but Dinero.js v2 should also allow for `bigint` [as soon as the proposal goes to stage 4](https://tc39.es/proposal-bigint/).
+
+In the meantime, and to support environments with no `bigint` support, Dinero.js v2 should let users provide their own `Calculator`. This way, they can use whatever library they want and map the calculator functions. This behavior should be made available through a Dinero factory, so it doesn't "pollute" the API.
+
+```js
+import { createDinero } from "@dinero.js/helpers";
+import Big from "some-bigint-library";
+
+const Dinero = createDinero({
+  calculator: {
+    add: Big.add,
+    subtract: Big.subtract
+    // ...
+  }
+});
+```
+
+A `Calculator` type should be provided to ease the creation with TypeScript:
+
+```ts
+const BigIntCalculator: Calculator = {
+  add: Big.add,
+  subtract: Big.subtract
+  // ...
+};
+```
+
+#### Currency service
+
+Currently, users must specify the `precision` of their `Dinero` objects, which is set to `2` by default. When a user creates objects with a currency which exponent is different from `2` (e.g., Iraqi dinars, Japanese yens, etc.), they must specify the `precision` by hand. One way to alleviate it is by setting a `defaultPrecision`, but this only works when one is using a single currency throughout their project.
+
+The currency service lets the user pass a `Currency` object instead of a plain string to the `currency` property when creating `Dinero` objects. `Currency` objects embark the currency's exponent, allowing the `Dinero` object to set the `precision` property automatically.
+
+```js
+import Dinero from "dinero.js";
+import { IQD } from "@dinero.js/currencies";
+
+// `precision` is set to `2` (default)
+const d1 = Dinero({ amount: 1000, currency: "IQD" });
+
+// `precision` is automatically set to `3`
+const d2 = Dinero({ amount: 1000, currency: IQD });
+```
+
+`Currency` objects could also embark the currency's symbol, allowing the library to fully own the currency formatting within the [`toFormat`](https://sarahdayan.github.io/dinero.js/module-Dinero.html#~toFormat) method and liberate itself from its dependency on the I18n API.
+
+```js
+import Dinero from "dinero.js";
+import { EUR } from "@dinero.js/currencies";
+
+const d1 = Dinero({ amount: 100, currency: "EUR" });
+const d2 = Dinero({ amount: 100, currency: EUR });
+
+d1.toFormat("$0,0.00"); // EUR100.00
+d2.toFormat("$0,0.00"); // €100.00
```

Comment:
Is there a way to specify whether you want the resulting string to use the EUR vs € in your example? It'd be nice to be able to choose in the format string itself whether you wanted the currency string or not I think.

Reply:
```
d1.toFormat("0,0.00  USD"); // 100.00 EUR
```
What would be even better is some way to say "use the exponent from the currency after the dot" so that you could use the same format for all currencies, regardless of the exponent they use. Moustache like substitutions come to mind, with exponent being converted to the correct number of zeros. This might be a clearer way to express it instead of having to use USD symbols, or maybe not.
```
d1.toFormat("0,0.{{exponent}} {{code}}"); // 100.00 EUR
```

Reply:
I've been iterating on that method, and ultimately I think the best solution is to let the user provide a render function:

```js
const price = Dinero({
  amount: 500,
  currency: EUR, // This is a `Currency` object, not a string as in v1
});

price.toFormat(({ amount, currency }) => `${currency.code} ${amount}`, {
  digits: 2,
  roundingMode: Math.round,
});

// "EUR 5"
```

The render function exposes `amount`, which is the human-readable amount which you get when invoking `toRoundedUnit`; and `currency`, the `Currency` object. The second argument are options passed to `toRoundedUnit`.

This allows for much more flexibility, and is a lot more expressive when reading the code.

If users want to keep using the I18n API, they can by writing their own adapter once:

```js
function i18n({ amount, currency }) {
  return amount.toLocaleString("en-GB", {
    style: "currency",
    currency: currency.code,
  });
}

price.toFormat(i18n, {
  digits: 2,
  roundingMode: Math.round,
});

// "€5.00"
```

Or they can do it more manually, like such:

```js
const currencySymbols = {
  EUR: "€",
  USD: "$",
  // ...
};

function manualI18n({ amount, currency }) {
  const { code } = currency;
  const symbol = currencySymbols[code] || code;

  return `${symbol}${amount.toFixed(2)}`;
}

price.toFormat(manualI18n, {
  digits: 2,
  roundingMode: Math.round,
});

// "€5.00"
```

It's a lot less confusing than the mask used in v1, and it doesn't couples the library to an API which isn't consistently implemented, nor fully supported in most environments.